### PR TITLE
[SPARK-46864][SS] Onboard Arbitrary StateV2 onto New Error Class Framework

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

This PR proposes to apply error class framework to the new data source, State API V2.

### Why are the changes needed?

Error class framework is a standard to represent all exceptions in Spark.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Refactored unit tests to check that the right error class was being thrown in certain situations

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by HeartSaVioR — COMMENTED** [review]
Starting with high level general comments, I haven't had a chance to review the whole change.

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -3383,6 +3401,18 @@
     ],
     "sqlState" : "428EK"
   },
+  "TWS_IMPLICIT_KEY_NOT_FOUND" : {
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:3404]: Yep. Also we have categories of internal errors, see
https://github.com/apache/spark/blob/f9f413e5ff6abe00a664e2dc75fb0ade2ff2986a/common/utils/src/main/scala/org/apache/spark/SparkException.scala#L99
You should add a suffix to your feature specific internal errors.

[the remark this answers is not in the document]
Comment (ericm-db) [reply common/utils/src/main/resources/error/error-classes.json:3404]: So do we still want to define this in `error-classes.json`, or just throw an InternalError from `StateStoreErrors`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MemoryStateStore.scala [hunk 2]
```diff
@@ -30,7 +30,7 @@ class MemoryStateStore extends StateStore() {
   }
 
   override def createColFamilyIfAbsent(colFamilyName: String): Unit = {
-    throw new UnsupportedOperationException("Creating multiple column families is not supported")
+    throw StateStoreErrors.multipleColumnFamilies("MemoryStateStoreProvider")
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MemoryStateStore.scala:33]: Is it possible to augment more tests to ensure that the NERF is being used correctly ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  }
+}
+
+class TransformWithStateMultipleColumnFamilies(stateStoreProvider: String)
+  extends SparkUnsupportedOperationException(
+    errorClass = "STAT_STORE_MULTIPLE_COLUMN_FAMILIES",
+    messageParameters = Map("stateStoreProvider" -> stateStoreProvider)
+  )
+
+// Used for ListState
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:60]: Lets remove these for now ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    new TransformWithStateUnsupportedOperation(operationName, entity)
+  }
+  def valueShouldBeNonNull(typeOfState: String): TransformWithStateValueShouldBeNonNull = {
+    new TransformWithStateValueShouldBeNonNull(typeOfState)
+  }
+}
+
+class TransformWithStateMultipleColumnFamilies(stateStoreProvider: String)
+  extends SparkUnsupportedOperationException(
+    errorClass = "STAT_STORE_MULTIPLE_COLUMN_FAMILIES",
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:56]: typo here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import org.apache.spark.{SparkException, SparkRuntimeException, SparkUnsupportedOperationException}
+
+/**
+ * Object for grouping error messages from (most) exceptions thrown from State API V2
+ *
+ * ERROR_CLASS has a prefix of "TWS_" representing State API V2.
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:25]: This needs to be updated ?

**Comment by anishshri-db** [comment]
@ericm-db - test failures also seem relevant. could you please take a look ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  )
+
+// Used for ListState
+class TransformWithStateMultipleValuesPerKey()
+  extends SparkRuntimeException(
+    errorClass = "STATE_STORE_STORE_MULTIPLE_VALUES_PER_KEY",
+    messageParameters = Map.empty
+  )
+
+class TransformWithStateUnsupportedOperation(operationType: String, entity: String)
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:67]: I thought it would be good to have a general unsupportedoperation exception

[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:67]: ok - maybe we can just add as needed ?

[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:67]: Just realized I use it here: https://github.com/apache/spark/pull/44883/files#diff-b7ff06fb7d0cd23d338c498a7119948e36377eff967a95c03bd449a66c815414R140

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala [hunk 7]
```diff
@@ -48,8 +48,7 @@ class ValueStateImpl[K, S](
   private def encodeKey(): UnsafeRow = {
     val keyOption = ImplicitGroupingKeyTracker.getImplicitKeyOption
     if (!keyOption.isDefined) {
-      throw new UnsupportedOperationException("Implicit key not found for operation on" +
-        s"stateName=$stateName")
+      throw StateStoreErrors.implicitKeyNotFound(stateName = stateName)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala:51]: We can just pass `stateName` directly ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 8]
```diff
@@ -256,8 +256,7 @@ private[sql] class HDFSBackedStateStoreProvider extends StateStoreProvider with
 
     // TODO: add support for multiple col families with HDFSBackedStateStoreProvider
     if (useColumnFamilies) {
-      throw new UnsupportedOperationException("Multiple column families are not supported with " +
-        s"HDFSBackedStateStoreProvider")
+      throw StateStoreErrors.multipleColumnFamilies("HDFSStateStoreProvider")
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:259]: Maybe rephrase as `multipleColumnFamiliesNotSupported` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  def implicitKeyNotFound(
+    stateName: String): SparkException = {
+    SparkException.internalError(
+      msg = s"Implicit key not found in state store for stateName=$stateName",
+      category = "TWS.IMPLICIT_KEY_NOT_FOUND"
+    )
+  }
+
+  def multipleColumnFamilies(stateStoreProvider: String):
+  TransformWithStateMultipleColumnFamilies = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:37]: Indent seems off ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    new TransformWithStateMultipleColumnFamilies(stateStoreProvider)
+  }
+
+  def unsupportedOperationException(operationName: String, entity: String):
+  TransformWithStateUnsupportedOperation = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:42]: same here ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 11]
```diff
@@ -167,4 +167,14 @@ class TransformWithStateValidationSuite extends StateStoreMetricsTest {
       }
     )
   }
+
+  test("transformWithState - creating colFamily with MemoryStateStore should fail") {
+    val stateStore = new MemoryStateStore()
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:172]: Could we just use HDFSBackedStateStoreProvider instead ?

File: docs/sql-error-conditions-internal-error-tws-error-class.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+[SQLSTATE: XX000](sql-error-conditions-sqlstates.html#class-XX-internal-error)
+
+Internal error using TransformWithStateOperator: 
+
+This error class has the following derived error classes:
+
+## IMPLICIT_KEY_NOT_FOUND
+
+`<message>`
+
```
Comment (anishshri-db) [inline docs/sql-error-conditions-internal-error-tws-error-class.md:36]: Nit: extra newline(s) ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  def multipleColumnFamiliesNotSupported(stateStoreProvider: String):
+    TransformWithStateMultipleColumnFamilies = {
+      new TransformWithStateMultipleColumnFamilies(stateStoreProvider)
+    }
+
+  def unsupportedOperationException(operationName: String, entity: String):
+    TransformWithStateUnsupportedOperation = {
+      new TransformWithStateUnsupportedOperation(operationName, entity)
+    }
+
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:45]: nit: extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+object StateStoreErrors {
+  def implicitKeyNotFound(
+    stateName: String): SparkException = {
+    SparkException.internalError(
+      msg = s"Implicit key not found in state store for stateName=$stateName",
+      category = "TWS.IMPLICIT_KEY_NOT_FOUND"
+    )
+  }
+
+  def multipleColumnFamiliesNotSupported(stateStoreProvider: String):
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:36]: Can we rephrase this to use StateStore prefix as well ?

`new StateStoreMultipleColumnFamiliesNotSupportedException` ?

Same for one below ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 15]
```diff
@@ -689,6 +689,20 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
 
+  testWithChangelogCheckpointingEnabled("RocksDB: Unsupported Operations" +
+    " with Changelog Checkpointing") {
+    val dfsRootDir = new File(Utils.createTempDir().getAbsolutePath + "/state/1/1")
+    val fileManager = new RocksDBFileManager(
+      dfsRootDir.getAbsolutePath, Utils.createTempDir(), new Configuration)
+    val changelogWriter = fileManager.getChangeLogWriter(1)
+
+    val ex = intercept[UnsupportedOperationException] {
+      changelogWriter.put("a", "1", "testColFamily")
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:700]: lets add a check for `delete` too ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala [hunk 16]
```diff
@@ -184,4 +187,19 @@ class ValueStateSuite extends SharedSparkSession
       assert(testState2.get() === null)
     }
   }
+
+
+  test("Creating colFamily with HDFS should fail") {
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:192]: nit: `colFamily with HDFSBackedStateStoreProvider should fail`

File: docs/sql-error-conditions-internal-error-tws-error-class.md (hunk 12 above)
Comment (ericm-db) [reply docs/sql-error-conditions-internal-error-tws-error-class.md:36]: This is a generated file

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala [hunk 17]
```diff
@@ -184,4 +187,19 @@ class ValueStateSuite extends SharedSparkSession
       assert(testState2.get() === null)
     }
   }
+
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:190]: nit: extra newline ?

**Review by anishshri-db — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json [hunk 18]
```diff
@@ -1656,6 +1656,19 @@
     ],
     "sqlState" : "XX000"
   },
+  "INTERNAL_ERROR_TWS" : {
+    "message" : [
+      "Internal error using TransformWithStateOperator: "
+    ],
+    "subClass" : {
+      "IMPLICIT_KEY_NOT_FOUND" : {
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:1664]: Internal errors should be not visible to users, so, don't need to introduce any sub-classes. Please, remove it and move `<message>` to `"Internal error using TransformWithStateOperator: "`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+ * Object for grouping error messages from (most) exceptions thrown from State API V2
+ *
+ * ERROR_CLASS has a prefix of "TWS_" or "STATE_STORE_" to indicate where the error is from
+ */
+object StateStoreErrors {
+  def implicitKeyNotFound(
+    stateName: String): SparkException = {
+    SparkException.internalError(
+      msg = s"Implicit key not found in state store for stateName=$stateName",
+      category = "TWS.IMPLICIT_KEY_NOT_FOUND"
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:32]: Please, don't mix `category` and error sub-class.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.spark.{SparkException, SparkUnsupportedOperationException}
+
+/**
+ * Object for grouping error messages from (most) exceptions thrown from State API V2
+ *
+ * ERROR_CLASS has a prefix of "TWS_" or "STATE_STORE_" to indicate where the error is from
+ */
+object StateStoreErrors {
+  def implicitKeyNotFound(
+    stateName: String): SparkException = {
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:29]:
```suggestion
  def implicitKeyNotFound(stateName: String): SparkException = {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import org.apache.spark.{SparkException, SparkUnsupportedOperationException}
+
+/**
+ * Object for grouping error messages from (most) exceptions thrown from State API V2
+ *
+ * ERROR_CLASS has a prefix of "TWS_" or "STATE_STORE_" to indicate where the error is from
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreErrors.scala:25]: Update the comment.

File: common/utils/src/main/resources/error/error-classes.json [hunk 22]
```diff
@@ -3230,6 +3243,24 @@
     ],
     "sqlState" : "0A000"
   },
+  "STATE_STORE_MULTIPLE_COLUMN_FAMILIES" : {
+    "message" : [
+      "Creating multiple column families with <stateStoreProvider> is not supported"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3248]:
```suggestion
      "Creating multiple column families with <stateStoreProvider> is not supported."
```

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:3248]: Cannot it be a sub-class of `UNSUPPORTED_FEATURE`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala [hunk 23]
```diff
@@ -184,4 +187,18 @@ class ValueStateSuite extends SharedSparkSession
...
+    val provider = new HDFSBackedStateStoreProvider()
+    val storeConf = new StateStoreConf(new SQLConf())
+    val ex = intercept[StateStoreMultipleColumnFamiliesNotSupportedException] {
+      provider.init(
+        storeId, keySchema, valueSchema, 0, useColumnFamilies = true,
+        storeConf, new Configuration)
+    }
+    assert(ex.getMessage.contains("Creating multiple column families" +
+      " with HDFSStateStoreProvider is not supported"))
+    assert(ex.getMessage.contains("[STATE_STORE_MULTIPLE_COLUMN_FAMILIES]"))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:202]: Please, use `checkError` here and other places.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: common/utils/src/main/resources/error/error-classes.json [hunk 24]
```diff
@@ -3230,6 +3236,18 @@
...
+    "message" : [
+      "Store does not support multiple values per key"
+    ],
+    "sqlState" : "42802"
+  },
+  "STATE_STORE_UNSUPPORTED_OPERATION" : {
+    "message" : [
+      "<operationType> operation not supported with <entity>"
+    ],
+    "sqlState" : "XXKST"
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-classes.json:3249]: Can a user get here? If so that would imply a 0A*** sqlstate rather than an internal error.

**Review by srielau — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json (hunk 24 above)
Comment (ericm-db) [reply common/utils/src/main/resources/error/error-classes.json:3249]: No, a user cannot get here. This can stay as an internal error.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
I'll leave this for today to see @MaxGekk can chime in - I'll merge this by tomorrow if he has no more comment by today. He can also approve and merge by today.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.
